pkg/stanza/operator/transformer/assign_keys/assign_keys.go

```diff
@@ -0,0 +1,132 @@
+func NewConfig() *Config {
+	return NewConfigWithID(operatorType)
+}
+
+// NewConfigWithID creates a new assign_keys operator config with default values
+func NewConfigWithID(operatorID string) *Config {
+	return &Config{
+		TransformerConfig: helper.NewTransformerConfig(operatorID, operatorType),
+	}
+}
+
+// Config is the configuration of a assign_keys operator
+type Config struct {
+	helper.TransformerConfig `mapstructure:",squash"`
+	Field                    entry.Field `mapstructure:"field"`
+	Keys                     []string    `mapstructure:"keys"`
+}
+
+// Build will build an assign_keys operator from the supplied configuration
+func (c Config) Build(logger *zap.SugaredLogger) (operator.Operator, error) {
+	transformerOperator, err := c.TransformerConfig.Build(logger)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(c.Keys) == 0 {
+		return nil, fmt.Errorf("assign_keys missing required field keys")
+	}
+
+	if e, ok := c.Field.FieldInterface.(entry.BodyField); ok {
+		return &Transformer[entry.BodyField]{
+			TransformerOperator: transformerOperator,
+			Field:               e,
+			Keys:                c.Keys,
+		}, nil
+	}
+
+	if e, ok := c.Field.FieldInterface.(entry.ResourceField); ok {
+		return &Transformer[entry.ResourceField]{
+			TransformerOperator: transformerOperator,
+			Field:               e,
+			Keys:                c.Keys,
+		}, nil
+	}
+
+	if e, ok := c.Field.FieldInterface.(entry.AttributeField); ok {
+		return &Transformer[entry.AttributeField]{
+			TransformerOperator: transformerOperator,
+			Field:               e,
+			Keys:                c.Keys,
+		}, nil
+	}
+
+	return nil, fmt.Errorf("invalid field type: %T", c.Field.FieldInterface)
+}
+
+// Transformer transforms a list in the entry field into a map. Each value is assigned a key from configuration keys
+type Transformer[T interface {
+	entry.BodyField | entry.ResourceField | entry.AttributeField
+	entry.FieldInterface
+}] struct {
+	helper.TransformerOperator
+	Field T
+	Keys  []string
+}
+
+// Process will process an entry with AssignKeys transformation.
+func (p *Transformer[T]) Process(ctx context.Context, entry *entry.Entry) error {
+	return p.ProcessWith(ctx, entry, p.Transform)
+}
+
+// Transform will apply AssignKeys to an entry
+func (p *Transformer[T]) Transform(entry *entry.Entry) error {
+	inputListInterface, ok := entry.Delete(p.Field)
+	if !ok {
+		// The field doesn't exist, so ignore it
+		return fmt.Errorf("apply assign_keys: field %s does not exist on entry", p.Field)
+	}
+
+	inputList, ok := inputListInterface.([]any)
+	if !ok {
+		return p.rollbackTransformAndGetError(entry, inputListInterface, fmt.Sprintf("apply assign_keys: couldn't convert field %s to []any", p.Field))
+	}
+	if len(inputList) != len(p.Keys) {
+		return p.rollbackTransformAndGetError(entry, inputListInterface, fmt.Sprintf("apply assign_keys: field %s contains %d values while expected keys are %s contain %d keys", p.Field, len(inputList), p.Keys, len(p.Keys)))
```

Comment:
I think we should document what happens when the number of fields doesn't match the number of keys.